target_chains/solana/programs/pyth_solana_receiver/src/lib.rs

```diff
@@ -0,0 +1,332 @@
+    pub fn post_accumulator_update_vaa(
+        ctx: Context<PostAccUpdateDataVaa>,
+        data: Vec<u8>, // accumulatorUpdateData {vaa, updates: [] }
+    ) -> Result<()> {
+        // verify accumulator update data header
+        let accumulator_update_data = AccumulatorUpdateData::try_from_slice(data.as_slice())
+            .map_err(|_| ProgramError::InvalidArgument)?;
+        match accumulator_update_data.proof {
+            Proof::WormholeMerkle { vaa, updates: _ } => {
+                let vaa: Vaa<&RawMessage> = serde_wormhole::from_slice(vaa.as_ref()).unwrap();
+                let (header, body): (Header, Body<&RawMessage>) = vaa.into();
+
+                let post_vaa_ix_data = PostVAAInstructionData {
+                    version:            header.version,
+                    guardian_set_index: header.guardian_set_index,
+                    timestamp:          body.timestamp,
+                    nonce:              body.nonce,
+                    emitter_chain:      body.emitter_chain.into(),
+                    emitter_address:    body.emitter_address.0,
+                    sequence:           body.sequence,
+                    consistency_level:  body.consistency_level,
+                    payload:            body.payload.to_vec(),
+                };
+                let post_vaa_ix = Instruction {
+                    program_id: ctx.accounts.wormhole_program.key(),
+                    accounts:   vec![
+                        AccountMeta::new_readonly(ctx.accounts.guardian_set.key(), false),
+                        AccountMeta::new_readonly(ctx.accounts.bridge_config.key(), false),
+                        AccountMeta::new_readonly(ctx.accounts.signature_set.key(), false),
+                        AccountMeta::new(ctx.accounts.vaa.key(), false),
+                        AccountMeta::new(ctx.accounts.payer.key(), true),
+                        AccountMeta::new_readonly(Clock::id(), false),
+                        AccountMeta::new_readonly(Rent::id(), false),
+                        AccountMeta::new_readonly(ctx.accounts.system_program.key(), false),
+                    ],
+                    data:       (POST_VAA, post_vaa_ix_data).try_to_vec()?,
+                };
+                invoke(&post_vaa_ix, ctx.accounts.to_account_infos().as_slice())?;
+            }
+        }
+        Ok(())
+    }
+
+    #[allow(unused_variables)]
+    pub fn post_updates(
+        ctx: Context<PostUpdates>,
+        vaa_hash: [u8; 32], // used for pda seeds
+        price_updates: Vec<Vec<u8>>,
+    ) -> Result<()> {
+        let vaa = &ctx.accounts.posted_vaa; // let posted_vaa_data = PostedVaaData::try_deserialize_unchecked(&mut &**vaa.try_borrow_data()?)?;
+        let wh_message = WormholeMessage::try_from_bytes(vaa.payload.as_slice())
+            .map_err(|_| ReceiverError::InvalidWormholeMessage)?;
+        msg!("constructed wh_message {:?}", wh_message);
```

Comment:
the checks for the message emitter need to be added here (and removed from the instruction constraints)

Reply:
can you elaborate on why it needs to be here vs. as an instruction constraint?

Reply:
i added the message emitter checks here and in the postVaa wrapper. although there's no downside to having it in both, i think it makes more sense being checked in the postAccumulatorUpdateVaa ix than here.

Reply:
I believe someone can post a vaa directly (by calling PostVAA themselves), then pass that VAA account to this function. If they do that, any security checks in this function are bypassed. (This is also why i think this function is kind of useless)

These checks should not be an instruction constraint because the data source emitter (chain id and address) is configurable via a governance message on all other chains. Using the instruction constraint is much less flexible than using a configuration parameter.

Also you need to check not only that the chain is correct, but also that the emitter address is correct.